### PR TITLE
[DSv4.1] Add FROST tail RoPE and group32 FP4/FP8 QDQ

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
DeepSeek indexer and window/DSpark consumers apply RoPE, round to BF16, then
quantize and dequantize back into BF16 storage. Add `RopeQDQInplace` and
`rope_qdq_inplace` on SM100 to perform that stage in one kernel.

| Format | Contiguous BF16 input | Group scale |
| --- | --- | --- |
| E2M1 FP4 | `[N,H,128]`, H=1/4/32 | UE8M0, group 32 |
| E4M3 FP8 | `[...,512]` | UE8M0, group 32 |

Both paths apply adjacent-pair RoPE to the last 64 channels, preserve the
intermediate BF16 rounding and signed zero, and mutate the input in place.
The prepared `backend="frost"` API supports dynamic lengths, explicit streams,
CUDA Graph replay and 16/4/2-byte-aligned storage. Execute performs no tensor
allocation, repacking, synchronization or compilation. Lazy exports,
dependency gates, documentation, tests and standalone benchmarks are included.
Triton >=3.7.0 is required. Packed cache storage and backward/STE are outside
this floating inference QDQ API.

Two independent B200 runs per format measured the actual FE API against the
fastest valid measured native chain for each shape and cache regime:

| Stage | GPU-stage speedup across both runs |
| --- | ---: |
| Indexer RoPE + FP4 QDQ | **1.49–1.79×** |
| Window/DSpark RoPE + FP8 QDQ | **1.38–1.79×** |

Coverage includes S4096/16384, TP1/TP8 queries, compression ratios 1/2,
B1/4 window KV and one-/five-token DSpark. All shapes won all eight paired
blocks in both runs, cache regimes and metrics. GPU event and Graph-replay
wall time are reported separately; `hot` means no explicit L2 eviction,
without a cache-residency claim. These are stage measurements, not model
throughput or eager Python execute latency.

Validation:

- All 29 focused API tests passed on SM100 and B200, covering dynamic lengths,
  pointer offsets, support guards, alias rejection, exact rounding boundaries
  and changed-input Graph replay.
- Benchmarks retain 14 native controls and require bitwise agreement across
  seven input generations for Frost. Controls exact on both timed distributions
  remain competitors even if a retained stress case fails. The fastest controls
  use FlashInfer RoPE with source or Torch-compiled QDQ, including a no-FTZ variant.
- Both B200 repetitions passed source, numerical, route, GPU-isolation and
  thermal audits. Focused formatting, syntax and SPDX checks passed.
- The regression on base `f32bcc0c0ec6243c788183d23d3e8a0f25efcec4` reused the
  pinned FE binding. Kernel/API/benchmark files match the measured versions;
  this is not a new timing run or a fresh C++ binding build.

The math follows
[DeepSeek-V4.1-Flash](https://huggingface.co/deepseek-ai/DeepSeek-V4.1-Flash/blob/dba1be0a40aa45a94ad051997016db3960a90277/inference/model.py)
and `kernel.py`. Fused kernels, Torch controls, unpack glue, harness and
fixtures are authored for this work. Source and FlashInfer baselines retain
their attribution. The no-FTZ control only appends `--ftz=false` to unchanged
FlashInfer sources; CUDA FP4/FP8 headers informed PTX operand order.
FlashInfer's scalar-scale fused FP8 API implements a different scaling contract.

Publication validation: rebased onto `develop` at
`adb963c01b6f65a0622efda48092fbc853dd3cf5`. Operation kernels, APIs,
benchmark controls and focused test files are byte-for-byte identical to the
validated versions described above. The rebase preserves the new upstream
lazy exports and documentation entries. Source identity, syntax, Black 26.3.1
(line length 160), SPDX headers and diff checks pass. Performance results above predate
this rebase.

Post-rebase host validation on these exact published heads now passes. A fresh
common C++ binding was built with GCC 11.4, CUDA 13.2.46, cuDNN 9.28.0 and
CPython 3.12.12; all binding sources, headers and CMake inputs are identical
across #1113–#1116. Each PR's byte-verified Python package passes actual import,
lazy-export and missing-dependency checks with that new binding. CPU metadata
checks also pass, with hardware queries explicitly mocked as SM100 and no GPU
kernel execution.

Fresh post-rebase B200 regression passed on 2026-09-17 at
`4016c9403dfc3a658376a59cfd62a3e3c027d1ea`: **29 focused API tests passed**,
with zero failures, errors or skips, using the fresh binding and cuDNN 9.28.0.
All 165 focused tests across #1113–#1116 passed on the same B200; loaded
package files and the binding were checked against their source hashes.
This adds correctness and integration validation. Performance figures above
still refer to the original measured versions; this regression did not time kernels.

Compatibility follow-up: move the QDQ API from `rope/api.py` to `rope/qdq.py` so it can coexist with the independent Vision RoPE backward API in #1122. The moved implementation is byte-identical; public names, signatures and kernel modules are unchanged. Benchmark provenance now hashes the new API path. The group16 and tail branches carry the same move.

The updated Python packages pass real import, lazy-export and CPU metadata checks with a freshly built current-develop FE binding (cuDNN 9.26 headers/backend). A local integration preview of all six DSv4.1 PRs also verifies that Vision, QDQ and tail resolve to separate modules and that their compile helpers cannot overwrite one another. Hardware queries are explicitly mocked for these metadata checks; this follow-up adds no GPU execution or timing results.

The QDQ cache regression now imports `cudnn.rope.qdq`, so it observes the QDQ compiler rather than the Vision compiler in the combined package. This follow-up changes only that test import.

Combined GPU integration validation: a local merge preview of #1113, #1114, #1115, #1116, #1122 and #1179 passed **139 focused B200 tests with zero failures, skips or deselections** in one package: Engram 22, group32/group16 RoPE QDQ 53, BF16 tail RoPE 32, Vision backward 29, and mHC projection backward 3. This includes the >2^31-element tail case, changed-input graph replay, explicit streams, numerical references and the mHC two-kernel route check. The test process exited 0; all loaded FE files and the freshly built binding matched the frozen sources. Runtime versions were cuDNN 9.28, PyTorch 2.13/cu130, CuTe DSL 4.7, Triton 3.7.1 and cuda-tile 1.5. The binding was compiled from current-develop sources against cuDNN 9.26 headers.

The enclosing validation wrapper subsequently exited nonzero because its compiler-cache footprint exceeded a 512 MiB bookkeeping limit (about 637 MiB); that failure and the complete test artifacts are retained separately. This run adds combined-package correctness evidence, with no new performance measurements.

## Summary by CodeRabbit

- **New Features**
  - Added experimental SM100 frontend APIs for in-place tail RoPE with microscaled FP4 or FP8 quantization.
  - Supports prepared operations, dynamic token counts, alignment options, CUDA streams, and repeated execution.
  - Added validation for tensor layouts, devices, shapes, aliasing, gradients, and hardware compatibility.

- **Documentation**
  - Added API reference, usage guidance, navigation links, and benchmark documentation.

- **Tests**
  - Added coverage for exact numerical results, edge cases, CUDA graphs, model-shaped workloads, and runtime validation.

- **Benchmarking**
  - Added FP4/FP8 performance benchmarks, reference implementations, result auditing, and baseline comparisons.

### CODE REVIEW

File: benchmark/rope_qdq/bench_fp4.py [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+                        assert not metrics["passed"]
+                for role in ROLES:
+                    for mode in ("eager", "graph"):
+                        reset(values)
+                        invoke(role) if mode == "eager" else graphs[role].replay()
+                        for index, actual in enumerate(buffers):
+                            key = f"{tag}/{name}/{role}/{mode}/{index}"
+                            metrics = compare(actual, expected)
+                            assert bool((storage[index][:64] == 123).all() and (storage[index][-64:] == 123).all())
+                            metrics["guards_unchanged"] = True
```
[the remark this answers is not in the document]
Comment (YangXu1990uiuc) [reply benchmark/rope_qdq/bench_fp4.py]: Fixed in 22f075b2bc35. Both FP4 and FP8 benchmarks now persist the computed `guards_unchanged` value before raising an explicit `RuntimeError`, so disabling assertions cannot hide a corrupted guard. CPU probes of the actual check blocks cover intact guards and corruption at either end under normal and optimized compilation; the old code fails these negative controls and the fix passes. Kernel/API code is unchanged; no new GPU timing is claimed.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `14de2846757e` against base `9e9050c4013e`.

No approval-blocking kernel/API issue found. The prepared ABI, exact rounding, alignment variants and replay behavior held up, and the benchmark guard-recording fix is independently confirmed. One nonblocking source/audit-gate follow-up remains. This is comment-only because the review uses the author's account.

**[P2] Keep source-identity and audit gates effective under optimized Python** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/14de2846757e8d452c06f0198f3d6e2aceec2991/benchmark/rope_qdq/model_contract.py#L52))

The guard-persistence fix is correct, but source pinning still relies on assertions. With `python -O` / `PYTHONOPTIMIZE`, verify_source accepts changed model.py/kernel.py/config.json and reports the pinned revision; audit_result.py likewise loses its assertion-based numerical/hash/route checks. Please use explicit exceptions for admission gates, or fail early when assertions are disabled. This is the parent-side counterpart of the nonblocking benchmark comment on #1115; it does not dispute ordinary-mode numerical results.

Evidence: Both fp4 and fp8 verify_source paths reject altered structural fixtures with optimize=0 and accept the same files with optimize=2, returning revision dba1be0a40aa45a94ad051997016db3960a90277. The probe parses source only and executes no supplied model code. Attribution: Introduced with this benchmark's model contract and assertion-based auditor.

Validation: Reviewed the API/kernel, exports/docs/tests, both complete FP4/FP8 benchmark paths, Torch/native controls, source contract, cache setup and auditor. The benchmarks restore disjoint inputs outside timing, keep stress-failing controls eligible when exact on the actual timed distributions, and include native unpack/writeback costs. Verified byte-identical QDQ API, kernel and focused tests against the recently tested #1116 head b6bc4f725ef6: all 29 parent QDQ cases passed within that 61-test run on SM100. Coverage includes int32/int64 positions, dynamic lengths, 16/4/2-byte alignment, alias rejection, exact midpoints/signed zeros, no tensor work on execute and changed-input explicit-stream graph replay. This reuses proven source-identical coverage rather than claiming a second 29-test execution. Additional exact-head SM100 probe: FP4 and FP8 constructed group patterns from all 65,280 finite BF16 bit patterns, mixing signs and signed zeros, and matched the reference bit-for-bit across 8,355,840 and 33,423,360 output elements respectively. Expected overflow at the largest BF16 magnitudes also matched the reference. Both saved BF16-boundary fixtures passed; deliberately omitting the intermediate BF16 rounding failed each control. Exact-head performance evidence reused from the parent leg of the #1115 comparison: two runs over FP4 [4,32,128]/[8192,32,128] and FP8 [4,512]/[32768,512] on a local 68-SM SM100. Prepared CPU enqueue was 16.39–17.25 us; GPU graph time approximately 1.88/95.6/1.91/45.75 us respectively. Thirty disjoint restored buffers, bitwise reference checks and changed-input replay were verified. These are current-head component timings, not B200 speedups. The CodeRabbit guard-recording issue is resolved. Extracted the actual check blocks from both benchmarks and tested intact/front-corrupted/back-corrupted guards with optimize=0 and optimize=2: the head persists false and raises, while the old optimized code records true and accepts corruption. The prior nonfinite-diagnostic JSON fix was also confirmed on byte-identical benchmark functions during #1115. Style/Docs and git diff --check pass.

Limitations: The full native-baseline benchmark sweeps, independent B200 repetitions and combined-package 139-test run remain author-provided evidence; they were not rerun here. The local reference/finite-pattern probes establish their stated cases, not model throughput or every possible RoPE cache/group combination. No extra broad CI was triggered: source-identical focused GPU coverage and exact-head boundary/performance probes are already available, along with the owner's recent combined integration run.

Comment only: GitHub does not allow reviewing one's own PR with approval/request-changes.

**Comment by YangXu1990uiuc** [comment]
Addressed the benchmark source/admission comment in d69d09e7f. Both benchmark entrypoints, `model_contract.py` and `audit_result.py` now explicitly reject optimized Python at module entry. The README documents running without `-O`, `-OO`, or `PYTHONOPTIMIZE`, keeping all existing source, numerical and audit assertions effective.

Reproduced the original issue with changed structural fixtures: normal Python rejects them, while optimized FP4/FP8 verification reported the pinned revision. An invalid numerical metric was also accepted by the optimized auditor. The fixed modules reject optimized imports before those gates can be removed; normal-mode rejection and benchmark `--help` still pass. The combined parent/compressed-KV verification contains 22 passing CPU controls. AST comparison verifies the original normal-mode computations/admission logic is unchanged aside from the explicit entry guard. Changed-file pre-commit passes. Kernel/API/test sources are unchanged; no GPU retiming is claimed.

The shared fix has also been merged into dependent #1115 and #1116 with ordinary fast-forward pushes.

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `d69d09e7fd9d` against base `9e9050c4013e`.

The remaining benchmark source/admission finding is resolved. Both benchmark entry points, the source contract and the auditor now fail explicitly under optimized Python, preserving their assertion-based checks in supported runs. No remaining findings in this follow-up. Comment-only because the posting account owns the PR.

Validation: All 20 PR-specific files were compared with reviewed 14de2846757e: only four benchmark/source/auditor modules and their README changed. API, device kernels and GPU tests remain byte-identical. All five changed files are byte-identical to the shared fix independently verified on #1115 at dad69aa61a24. Reused those fresh CPU results: all four modules reject -O and -OO before imports, both benchmark help commands and auditor help work normally, and ordinary-mode source verification rejects changed bytes. AST comparison confirms ordinary-mode executable logic is unchanged apart from the entry guards. Reused prior source-identical QDQ GPU evidence: 29 focused tests plus exhaustive finite-BF16 group-pattern probes and existing-path performance checks. No new GPU execution or timing is claimed for this benchmark-only follow-up. Current-head pre-commit/fern and whitespace checks pass.

Limitations: The full B200 benchmark sweeps remain author-provided evidence, with the scope stated in the earlier review. No additional broad CI requested for this small benchmark-only fix.

Comment only: GitHub does not allow reviewing one's own PR with approval/request-changes.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss python_tests

**Review by vedaanta — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `d69d09e7fd9d` against base `774c6e6c51a4`.

Independently reviewed the group32 FP4/FP8 implementation, prepared API, exports, tests and benchmark/auditor paths. No actionable correctness finding. The intermediate BF16 rounding, signed-zero handling, dynamic row ABI and runtime layout/alias guards are consistent with the documented contract. The prior benchmark assertion/guard-recording comments are resolved.

Validation: Exact-head SM100 run: 29 passed, zero skipped, with explicit L0/L1 selection. Covers both position dtypes, dynamic lengths, 16/4/2-byte alignment, exact midpoints/signed zeros, aliases, model-shaped cases, explicit-stream changed-input graph replay and no tensor work/compilation during execute. Torch 2.13.0+cu130, Triton 3.7.1, cuDNN 9.26; verified checkout imports and byte-identical native sources for the reused binding. Both saved FP4/FP8 BF16-boundary fixtures passed; intentionally omitting intermediate BF16 rounding failed both negative controls. Eight optimized-Python benchmark/source/auditor rejection checks passed. Changed-file whitespace check passed. Small current-head component probe used 16 disjoint buffers per graph and checked every output bitwise. FP4 shapes [4,4,128]/[4096,4,128] measured approximately 2.00/7.99 us GPU; FP8 [4,512]/[4096,512] approximately 2.26/7.85 us. CPU enqueue medians ranged 19-38 us. These are local SM100 component observations, not a speedup comparison. Existing current-head GPU CI is running; no duplicate pipeline requested.

Limitations: The full FlashInfer/source baseline sweeps and B200 speedups remain author-reported; not independently reproduced here. Position values and finite inputs remain documented caller preconditions. CI is pending.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
Resolved the develop conflict after #1113 by preserving both the Engram and RoPE lazy exports/install hints. Current head: `dd8c7650f9b2041601a65869215062de2bec7dc8`. All QDQ operation sources are byte-identical to the approved and CI-tested head. Exact-package import/lazy-export/missing-dependency checks passed for both families without initializing CUDA; file-scoped pre-commit and the PR diff check passed. No duplicate GPU CI run was requested.
